server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -78,393 +90,41 @@ public HealthIndicatorResult calculate(boolean explain, HealthInfo healthInfo) {
-            .collect(Collectors.toSet());
-    }
-
-    // Non-private for unit testing
-    static Set<DiscoveryNodeRole> getRolesOnNodes(Set<String> nodeIds, ClusterState clusterState) {
-        return clusterState.nodes()
-            .getNodes()
-            .values()
-            .stream()
-            .filter(node -> nodeIds.contains(node.getId()))
-            .map(DiscoveryNode::getRoles)
-            .flatMap(Collection::stream)
-            .collect(Collectors.toSet());
-    }
-
-    // Non-private for unit testing
-    static Set<String> getNodesWithDataRole(Set<String> nodeIds, ClusterState clusterState) {
-        return clusterState.nodes()
-            .getNodes()
-            .values()
-            .stream()
-            .filter(node -> nodeIds.contains(node.getId()))
-            .filter(node -> node.getRoles().stream().anyMatch(DiscoveryNodeRole::canContainData))
-            .map(DiscoveryNode::getId)
-            .collect(Collectors.toSet());
-    }
-
-    // Non-private for unit testing
-    static Set<String> getNodesWithMasterRole(Set<String> nodeIds, ClusterState clusterState) {
-        return clusterState.nodes()
-            .getNodes()
-            .values()
-            .stream()
-            .filter(node -> nodeIds.contains(node.getId()))
-            .filter(node -> node.canContainData() == false && node.isMasterNode())
-            .map(DiscoveryNode::getId)
-            .collect(Collectors.toSet());
-    }
-
-    // Non-private for unit testing
-    static Set<String> getNodesWithNonDataNonMasterRoles(Set<String> nodeIds, ClusterState clusterState) {
-        return clusterState.nodes()
-            .getNodes()
-            .values()
-            .stream()
-            .filter(node -> nodeIds.contains(node.getId()))
-            .filter(node -> node.canContainData() == false && node.isMasterNode() == false)
-            .map(DiscoveryNode::getId)
-            .collect(Collectors.toSet());
-    }
-
-    // Non-private for unit testing
-    static Set<String> getIndicesForNodes(Set<String> nodes, ClusterState clusterState) {
-        return clusterState.routingTable()
-            .allShards()
-            .stream()
-            .filter(routing -> nodes.contains(routing.currentNodeId()))
-            .map(routing -> routing.index().getName())
-            .collect(Collectors.toSet());
-    }
+        HealthIndicatorDetails details = getDetails(explain, diskHealthInfoMap, blockedIndices);
 
-    // Non-private for unit testing
-    static Set<String> getNodeIdsForIndices(Set<String> indices, ClusterState clusterState) {
-        return clusterState.routingTable()
-            .allShards()
-            .stream()
-            .filter(routing -> indices.contains(routing.index().getName()))
-            .map(ShardRouting::currentNodeId)
-            .collect(Collectors.toSet());
+        DiskHealthAnalyzer diskHealthAnalyzer = new DiskHealthAnalyzer(diskHealthInfoMap, blockedIndices, clusterState);
```

Comment:
Since the `DiskHealthAnalyzer` already has the cluster state, could it look and compute the blocked indices? 

We now compute them in the indicator but not use them anywhere else except passing them to `DiskHealthAnalyzer`.

What do you think?

Reply:
That is true, my reasoning was that the interface shows the information needed to calculate it, but we can document why we need the cluster state and leave it at that ;) .

Reply:
in that case I will move the details too.